React/training/src/redux/Game/reducer.js

```diff
@@ -0,0 +1,57 @@
+const initialState = {
+  history: [
+    {
+      squares: Array(9).fill(null)
+    }
+  ],
+  stepNumber: 0,
+  xIsNext: true,
+  winner: null
+}
+
+function reducer (state = initialState, action) {
```

Comment:
Here you have a lot of cases that are not neccesary.
I think that this cases are enough: 
```
  switch (action.type) {
    case actions.SQUARE_CLICKED:
      return {
        ...state,
        history: history.concat([
          {
            squares
          }
        ]),
        xIsNext: !state.xIsNext,
        stepNumber: history.length,
        winner
      };
    case actions.HISTORY_ITEM_SELECTED:
      return {
        ...state,
        stepNumber: action.index,
        xIsNext: action.index % 2 === 0,
        winner,
        history: state.history.slice(0, action.index + 1)
      };
    default:
      return state;
  }
```